### PR TITLE
Improve chromeext annotations

### PR DESCRIPTION
**Pull request by attfarhan**
WIP. Fix bugs related to chrome extension annotations.  

Implemented: 
- Adds refreshVCS and automatically getting the latest annotations.  
- Fixes bug where jump-to-def will automatically revert to master branch.  Jump-to-def will now stay within the currently selected branch. 
- Adds revision to the link to trigger builds when search 404s
- Adds checks to fix syntax highlighting issues when trying to solve the split up annotations in go imports because of quote mark span tags.
- Adds checks to ensure '<span>' appearing anywhere in comments or in the code won't break annotations. 

##### Reviewer tasks
- [ ] HACKS: reviewer approves hacks introduced by this change

##### Test plan
Build this locally and test the refresh and annotations update by pushing to a repo and seeing that broken annotations are not added.  Once the build on Sourcegraph is finished, see that correct annotations are added. 

Try jump-to-def on a branch that isn't master. https://github.com/alexo/wro4j/blob/1.7.x/wro4j-core/src/main/java/ro/isdc/wro/http/WroFilter.java#sourcegraph&def=JavaArtifact/ro.isdc.wro4j/wro4j-core/-/ro/isdc/wro/http/WroFilter:type/createConfiguration&L135-143 

View code that has <span> comments and see that anns don't break. https://github.com/attfarhan/mux/blob/master/route.go (build for this is queued, but should work when it's finished).  

Click on an external def from a branch that isn't master and make sure we don't 404, and we get to the def on the default branch of the repository.

### CODE REVIEW

**Comment by sqs** [comment]
@attfarhan Because the event listener is listening on click events on ALL child elements of document (https://github.com/sourcegraph/sourcegraph/blob/37d7727760b8a384f690c8bf1593c843c6c32b88/client/browser-ext/chrome/extension/inject.js#L80), `ev.target` and `ev.currentTarget` will be the deepest node that was clicked (basically).

So now that there can be spans inside the links, we need to do some node traversal.

Specifically, at https://github.com/sourcegraph/sourcegraph/blob/37d7727760b8a384f690c8bf1593c843c6c32b88/client/browser-ext/chrome/extension/inject.js#L110, we need to have some code that tries to climb the DOM tree to look for the `data-sourcegraph-ref` element. But since the deepest nesting underneath a `data-sourcegraph-ref` element is a single level, we can just look to see whether the event target or its parent has `data-sourcegraph-ref`; we don't need to iterate. So this should work:

```
	// ...
	_clickRef(ev) {
		if (typeof ev.target.dataset.sourcegraphRef !== "undefined" || (ev.target.parentNode && typeof ev.target.parentNode.dataset.sourcegraphRef !== "undefined")) {
	// ...
```

**Comment by attfarhan** [comment]
This has been updated.  @sqs I didn't add the case above since I changed the way I handled the issue.

File: client/browser-ext/chrome/extension/inject.js [hunk 1]
```diff
@@ -132,7 +132,9 @@ class InjectApp extends React.Component {
 		const urlsplit = loc.pathname.slice(1).split("/");
 		let user = urlsplit[0];
 		let repo = urlsplit[1]
-		let rev = "master"
+		// We scrape the current branch and set rev to it so we stay on the same branch when doing jump-to-def
+		let currBranch = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0] ? document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0].title : "master"
```
Comment (sqs) [inline client/browser-ext/chrome/extension/inject.js]: Extract the `document.getElementsByClassName('select-menu-button js-menu-target css-truncate')` expression into a var, since it's quite long.

File: client/browser-ext/chrome/extension/inject.js [hunk 2]
```diff
@@ -173,8 +175,19 @@ class InjectApp extends React.Component {
 		this.addSearchButton();
 
 		let {user, repo, rev, path, defPath} = this.parseURL();
+		// This scrapes the latest commit ID and updates rev to the latest commit so we are never injecting
+		// outdated annotations.  If there is a new commit, srclib-data-version will return a 404, but the
+		// refresh endpoint will update the version and the annotations will be up to date once the new build succeeds
+		let latestRev = document.getElementsByClassName("js-permalink-shortcut")[0] ? document.getElementsByClassName("js-permalink-shortcut")[0].href.split("/")[6] : rev
+		// TODO: Branches that are not built on Sourcegraph will not get annotations, need to trigger
+		let currBranch = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0] ? document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0].title : "master"
```
Comment (sqs) [inline client/browser-ext/chrome/extension/inject.js]: It's dangerous to ever hardcode "master". If a repo's default branch is not master, then the hardcoding will always be incorrect. In what cases can we not determine the branch name and need to fall back to "master"?

File: client/browser-ext/chrome/extension/inject.js [hunk 3]
```diff
@@ -132,7 +132,9 @@ class InjectApp extends React.Component {
 		const urlsplit = loc.pathname.slice(1).split("/");
 		let user = urlsplit[0];
 		let repo = urlsplit[1]
-		let rev = "master"
+		// We scrape the current branch and set rev to it so we stay on the same branch when doing jump-to-def
+		let currBranch = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0] ? document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0].title : "master"
+		let rev = currBranch
```
Comment (sqs) [inline client/browser-ext/chrome/extension/inject.js]: semicolon

**Comment by sqs** [comment]
Made a few comments. Overall looks great! Please get @rothfels to run through the test plan tomorrow before updating.

File: client/browser-ext/chrome/extension/inject.js [hunk 4]
```diff
@@ -132,7 +132,10 @@ class InjectApp extends React.Component {
 		const urlsplit = loc.pathname.slice(1).split("/");
 		let user = urlsplit[0];
 		let repo = urlsplit[1]
-		let rev = "master"
+		// We scrape the current branch and set rev to it so we stay on the same branch when doing jump-to-def
+		let branchSelectorButton = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0]
+		let currBranch = branchSelectorButton ? branchSelectorButton.title : "master";
```
Comment (attfarhan) [inline client/browser-ext/chrome/extension/inject.js]: @sqs I put "master" as a placeholder for when the user is is not on a blob or repo view on GitHub.  It'll show an error because it can't find property title of undefined, and this method is called in refreshState, which is called on every page movement.

**Comment by rothfels** [comment]
The "Find usages" link needs to carry the branch (it does not): http://cl.ly/1G0E3Z100K2X

**Comment by rothfels** [comment]
The first step of the test plan is broken, after I add new code: http://cl.ly/3x3e1F2M3v1c

**Comment by rothfels** [comment]
Hmm...never mind on the latest comment, I think I was running two versions of the chrome extension simultaneously.

**Comment by rothfels** [comment]
Ok @attfarhan test plan looks good, let's just update the find usages link to carry branch name.

File: client/browser-ext/app/containers/App.js [hunk 5]
```diff
@@ -110,7 +110,7 @@ export default class App extends Component {
 							<h3 styleName="list-item-empty">Searching...</h3>
 						}
 						{srclibDataVersionFetch && srclibDataVersionFetch.response && srclibDataVersionFetch.response.status === 404 &&
-							<h3 styleName="list-item-empty">404 Not Found: This repository has not been indexed by Sourcegraph.<br/><a href={`https://sourcegraph.com/${this.props.repo}`}>Let Sourcegraph index this repository.</a></h3>
+							<h3 styleName="list-item-empty">404 Not Found: This repository has not been indexed by Sourcegraph.<br/><a href={`https://sourcegraph.com/${this.props.repo}@${this.props.rev}`}>Let Sourcegraph index this repository.</a></h3>
```
Comment (rothfels) [inline client/browser-ext/app/containers/App.js]: Remove the @rev here, I don't think it's necessary.

File: client/browser-ext/chrome/extension/annotations.js [hunk 6]
```diff
@@ -74,13 +74,25 @@ function traverseDOM(annsByStartByte, annsByEndByte){
 			if (children[j].nodeType === Node.TEXT_NODE){
 				childNodeChars = children[j].nodeValue.split("");
 			} else {
+				// Quotation marks on GitHub are given their own span tags.
+				// This messes up anns for go imports. We handle this
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Not just for go imports, for all kinds of things, right?

File: client/browser-ext/chrome/extension/annotations.js [hunk 7]
```diff
@@ -74,13 +74,25 @@ function traverseDOM(annsByStartByte, annsByEndByte){
...
+				// with "pl-pds", the class for quotes, and there aren't any other
+				// tags between them (tags in between needed for highlighting).
 				if (children[j].children.length > 1) {
-					// HACK: combine children spans, e.g. for a quoted token
-					// there may be 3 spans, one for each quote and one
-					// for the token iteself
-					children[j].innerHTML = _.escape(children[j].innerText);
+					let goImport = true;
+					for (let k = 0; k < children[j].children.length; k++) {
+						if(children[j].children[k].className !== "pl-pds") {
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Space after `if`

File: client/browser-ext/chrome/extension/annotations.js [hunk 8]
```diff
@@ -74,13 +74,25 @@ function traverseDOM(annsByStartByte, annsByEndByte){
...
+							goImport = false
+							break;
+						}
+					}
+					if (goImport) {
+						children[j].innerHTML = _.escape(children[j].innerText);
+					}
 				}
-				childNodeChars = _.unescape(children[j].outerHTML).split("");
+				childNodeChars = (children[j].outerHTML).split("");
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Remove parens around `children[j].outerHTML` (unnecessary)

File: client/browser-ext/chrome/extension/annotations.js [hunk 9]
```diff
@@ -91,14 +103,21 @@ function traverseDOM(annsByStartByte, annsByEndByte){
...
 				}
-
 				if (!consumingSpan){
-					output += next(childNodeChars[k], startByte, annsByStartByte, annsByEndByte)
-					startByte += utf8.encode(childNodeChars[k]).length
-				}
-				else {
+					// Case to handle if < or > appears, so that we don't "consume" or make span tags in the code disappear
+					// This will not break if "&lt;" or "&gt;" appear because chars are escaped.
+					if (childNodeChars[k] === "&" && (((childNodeChars.slice(k, k+4).join("")) === ("&gt;")) || (childNodeChars.slice(k, k+4).join("") === ("&lt;")))) {
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: The conditional is long and hard to parse. Can you decompose it into a private helper function?

File: client/browser-ext/app/containers/App.js (hunk 5 above)
Comment (sqs) [inline client/browser-ext/app/containers/App.js]: Going to the rev will trigger a build for that rev so it is necessary. But the text should make it clear that that specific revision/branch hasn't been indexed, not the whole repo.

File: client/browser-ext/chrome/extension/inject.js (hunk 4 above)
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: @attfarhan isn't the current branch state always in the URL? Why do we have to parse it from the rev selector dom element state? Is it because there can be an unspecified revision in the URL for the default branch? If so, a comment here would help.

File: client/browser-ext/chrome/extension/inject.js [hunk 10]
```diff
@@ -132,7 +132,10 @@ class InjectApp extends React.Component {
 		const urlsplit = loc.pathname.slice(1).split("/");
 		let user = urlsplit[0];
 		let repo = urlsplit[1]
-		let rev = "master"
+		// We scrape the current branch and set rev to it so we stay on the same branch when doing jump-to-def
+		let branchSelectorButton = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0]
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: Use double quotes for all strings, like we have everywhere else.

File: client/browser-ext/chrome/extension/inject.js [hunk 11]
```diff
@@ -173,8 +176,20 @@ class InjectApp extends React.Component {
...
 
 		let {user, repo, rev, path, defPath} = this.parseURL();
+		// This scrapes the latest commit ID and updates rev to the latest commit so we are never injecting
+		// outdated annotations.  If there is a new commit, srclib-data-version will return a 404, but the
+		// refresh endpoint will update the version and the annotations will be up to date once the new build succeeds
+		let latestRev = document.getElementsByClassName("js-permalink-shortcut")[0] ? document.getElementsByClassName("js-permalink-shortcut")[0].href.split("/")[6] : rev
+		// TODO: Branches that are not built on Sourcegraph will not get annotations, need to trigger
+		let branchSelectorButton = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0];
+		let currBranch = branchSelectorButton ? branchSelectorButton.title : rev;
+		if (rev !== latestRev) rev = latestRev;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: I don't think you need the conditional `if`, do you?

File: client/browser-ext/chrome/extension/inject.js [hunk 12]
```diff
@@ -173,8 +176,20 @@ class InjectApp extends React.Component {
 		this.addSearchButton();
 
 		let {user, repo, rev, path, defPath} = this.parseURL();
+		// This scrapes the latest commit ID and updates rev to the latest commit so we are never injecting
+		// outdated annotations.  If there is a new commit, srclib-data-version will return a 404, but the
+		// refresh endpoint will update the version and the annotations will be up to date once the new build succeeds
+		let latestRev = document.getElementsByClassName("js-permalink-shortcut")[0] ? document.getElementsByClassName("js-permalink-shortcut")[0].href.split("/")[6] : rev
+		// TODO: Branches that are not built on Sourcegraph will not get annotations, need to trigger
+		let branchSelectorButton = document.getElementsByClassName('select-menu-button js-menu-target css-truncate')[0];
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: You have this code in two places; decompose into a function.

File: client/browser-ext/chrome/extension/inject.js [hunk 13]
```diff
@@ -173,8 +176,20 @@ class InjectApp extends React.Component {
...
+		let currBranch = branchSelectorButton ? branchSelectorButton.title : rev;
+		if (rev !== latestRev) rev = latestRev;
+		if (currBranch !== "master") rev = currBranch;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: I'm confused by what all this code is supposed to be doing. Can you explain to me in person?

File: client/browser-ext/app/containers/App.js (hunk 5 above)
Comment (rothfels) [inline client/browser-ext/app/containers/App.js]: Ahh, got it. Thanks.

**Comment by attfarhan** [comment]
@rothfels, this is updated.

File: client/browser-ext/app/containers/App.js (hunk 5 above)
Comment (rothfels) [inline client/browser-ext/app/containers/App.js]: @attfarhan can you clarify the copy here? `This repository (or revision) has not been indexed by Sourcegraph`

File: client/browser-ext/chrome/extension/annotations.js [hunk 14]
```diff
@@ -119,13 +137,14 @@ function traverseDOM(annsByStartByte, annsByEndByte){
 	}
 }
 
+function isSpanTagString(childNodeChars, k) {
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: k is a bit contextless here -- what is k?

I would change the name to `isStartOfSpanTag` and name the second arg `idx`

File: client/browser-ext/chrome/extension/annotations.js [hunk 15]
```diff
@@ -119,13 +137,14 @@ function traverseDOM(annsByStartByte, annsByEndByte){
...
+	return (childNodeChars[k] === "&" && (((childNodeChars.slice(k, k+4).join("")) === ("&gt;")) || (childNodeChars.slice(k, k+4).join("") === ("&lt;"))))
+}
+
 // next is a helper method for traverseDOM which transforms a character
 // into itself or wraps the character in a starting/ending anchor tag
 function next(c, byteCount, annsByStartByte, annsByEndByte) {
 	let matchDetails = annsByStartByte[byteCount];
-
-	c = _.escape(c); // IMPORTANT: escape all markup injected in HTML
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Why did this get removed?

File: client/browser-ext/chrome/extension/annotations.js [hunk 16]
```diff
@@ -140,20 +159,35 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
 	if (annotating && annsByEndByte[byteCount + 1]) {
 		annotating = false;
 		return `${c}</a>`;
 	}
-
 	return c;
 }
 
+export const defaultBranchCache = {};
+export const fetchingDefaultBranchCache = {};
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Can you leave a comment about what purpose this serves?

File: client/browser-ext/chrome/extension/annotations.js [hunk 17]
```diff
@@ -140,20 +159,35 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+function cacheDefaultBranch(annURL) {
+	let annURLsplit = [annURL.split("/")[1], annURL.split("/")[2], annURL.split("/")[3]];
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: A couple issues:
- You don't need to call `annURL.split("/")` 3 times; once will suffice
- What if the annURL doesn't have the form `github.com/user/repo/`? Maybe just leave a comment in here that we assume this is what structure the URL has so the next developer isn't confused. (And that if we can't fetch the default branch, we'll use "master" as a default so it's not the end of the world.

File: client/browser-ext/chrome/extension/annotations.js [hunk 18]
```diff
@@ -140,20 +159,35 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+	if (fetchingDefaultBranchCache[repo]) {
+		return;
+	}
+	fetchingDefaultBranchCache[repo] = true;
+	fetch(`https://sourcegraph.com/.api/repos/${repo}`)
+		.then((response) => {
+			defaultBranchCache[repo] = response.DefaultBranch;
+			fetchingDefaultBranchCache[repo] = false;
+		})
+		.catch((err) => console.log("Error getting default branch"));
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: This is an unreachable code path; fetch will never throw an error, the only way to get here would be if the `.then` chain did a `throw err`.

File: client/browser-ext/chrome/extension/annotations.js [hunk 19]
```diff
@@ -140,20 +159,35 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+function cacheDefaultBranch(annURL) {
+	let annURLsplit = [annURL.split("/")[1], annURL.split("/")[2], annURL.split("/")[3]];
+	let repo = annURLsplit.join("/");
+	if (fetchingDefaultBranchCache[repo]) {
+		return;
+	}
+	fetchingDefaultBranchCache[repo] = true;
+	fetch(`https://sourcegraph.com/.api/repos/${repo}`)
+		.then((response) => {
+			defaultBranchCache[repo] = response.DefaultBranch;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: You only want to execute this logic if the status code is 200.

File: client/browser-ext/chrome/extension/annotations.js [hunk 20]
```diff
@@ -140,20 +159,35 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+			fetchingDefaultBranchCache[repo] = false;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: You always want to set this to false (regardless of the response status code).

File: client/browser-ext/chrome/extension/annotations.js (hunk 19 above)
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Do you want to just store "master" here if there's a non 200 status code?

File: client/browser-ext/chrome/extension/annotations.js [hunk 21]
```diff
@@ -140,20 +159,35 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+export const defaultBranchCache = {};
+export const fetchingDefaultBranchCache = {};
+function cacheDefaultBranch(annURL) {
+	let annURLsplit = [annURL.split("/")[1], annURL.split("/")[2], annURL.split("/")[3]];
+	let repo = annURLsplit.join("/");
+	if (fetchingDefaultBranchCache[repo]) {
+		return;
+	}
+	fetchingDefaultBranchCache[repo] = true;
+	fetch(`https://sourcegraph.com/.api/repos/${repo}`)
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: You're shouldn't fetch if the default branch has already been fetched (i.e. `defaultBranchCache[repo]` is defined).

File: client/browser-ext/chrome/extension/annotations.js (hunk 18 above)
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Not saying you should make it reachable, just that you should never check in code that is unreachable. :)

Besides, you don't need to log the error, since chrome will automatically log network request errors (non-20x status codes)

File: client/browser-ext/chrome/extension/inject.js [hunk 22]
```diff
@@ -111,10 +113,13 @@ class InjectApp extends React.Component {
 		if (typeof ev.target.dataset.sourcegraphRef !== "undefined") {
 			let urlProps = this.parseURL({pathname: ev.target.pathname, hash: ev.target.hash});
 			urlProps.repo = `github.com/${urlProps.user}/${urlProps.repo}`;
+			// HACK: We need the default branch for external defs. This gets it for us, but we'd
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: I don't think this comment is relevant or necessary, you can remove it.

File: client/browser-ext/chrome/extension/inject.js [hunk 23]
```diff
@@ -132,7 +137,9 @@ class InjectApp extends React.Component {
 		const urlsplit = loc.pathname.slice(1).split("/");
 		let user = urlsplit[0];
 		let repo = urlsplit[1]
-		let rev = "master"
+		// We scrape the current branch and set rev to it so we stay on the same branch when doing jump-to-def
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: This is only true if we are if we are navigating within a repo, right? In which case, can't the rev be parsed out of the URL? And if we're navigating to an external repo, don't we want to use the `defaultBranchCache`?

File: client/browser-ext/chrome/extension/annotations.js (hunk 15 above)
Comment (attfarhan) [inline client/browser-ext/chrome/extension/annotations.js]: We no longer unescape in traverseDOM, we do this so we can read html in the code or comments as their character entities so we can make sure <span> tags in the code don't mess up the annotations.

File: client/browser-ext/chrome/extension/inject.js [hunk 24]
```diff
@@ -155,6 +162,10 @@ class InjectApp extends React.Component {
 		return info;
 	}
 
+	branchSelectorButton() {
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: Name this function `getBranchSelectorButton`.

File: client/browser-ext/chrome/extension/inject.js [hunk 25]
```diff
@@ -173,8 +184,17 @@ class InjectApp extends React.Component {
 		this.addSearchButton();
 
 		let {user, repo, rev, path, defPath} = this.parseURL();
+		// This scrapes the latest commit ID and updates rev to the latest commit so we are never injecting
+		// outdated annotations.  If there is a new commit, srclib-data-version will return a 404, but the
+		// refresh endpoint will update the version and the annotations will be up to date once the new build succeeds
+		let latestRev = document.getElementsByClassName("js-permalink-shortcut")[0] ? document.getElementsByClassName("js-permalink-shortcut")[0].href.split("/")[6] : rev;
+		// TODO: Branches that are not built on Sourcegraph will not get annotations, need to trigger
+		if (rev !== latestRev) rev = latestRev;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: Why is the `if` necessary -- does it add anything?

File: client/browser-ext/chrome/extension/inject.js [hunk 26]
```diff
@@ -173,8 +184,17 @@ class InjectApp extends React.Component {
 		this.addSearchButton();
 
 		let {user, repo, rev, path, defPath} = this.parseURL();
+		// This scrapes the latest commit ID and updates rev to the latest commit so we are never injecting
+		// outdated annotations.  If there is a new commit, srclib-data-version will return a 404, but the
+		// refresh endpoint will update the version and the annotations will be up to date once the new build succeeds
+		let latestRev = document.getElementsByClassName("js-permalink-shortcut")[0] ? document.getElementsByClassName("js-permalink-shortcut")[0].href.split("/")[6] : rev;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: Just making sure I understand this correctly -- is it replacing a branch name with a 64char CommitID?

File: client/browser-ext/chrome/extension/inject.js [hunk 27]
```diff
@@ -217,11 +237,12 @@ class InjectApp extends React.Component {
 	_directURLToDef({repo, rev, defPath, def}) {
 		const defObj = def ? def.content[keyFor(repo, rev, defPath)] : null;
 		if (defObj) {
-			const pathname = `/${repo.replace("github.com/", "")}/blob/${rev}/${defObj.File}`;
+			let defBranch = defaultBranchCache[repo]
+			const pathname = (repo === this.props.repo) ? `/${repo.replace("github.com/", "")}/blob/${rev}/${defObj.File}` : `/${repo.replace("github.com/", "")}/blob/${defBranch}/${defObj.File}`;
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: This ternary expression is long and hard to read. How about keeping the code as it was before, but subbing this line:

`if (repo !== this.props.repo) rev = defBranch || "master"`

File: client/browser-ext/chrome/extension/inject.js [hunk 28]
```diff
@@ -217,11 +237,12 @@ class InjectApp extends React.Component {
 	_directURLToDef({repo, rev, defPath, def}) {
 		const defObj = def ? def.content[keyFor(repo, rev, defPath)] : null;
 		if (defObj) {
-			const pathname = `/${repo.replace("github.com/", "")}/blob/${rev}/${defObj.File}`;
+			let defBranch = defaultBranchCache[repo]
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/inject.js]: You need a default, like `defaultBranchCache[repo] || "master"`

File: client/browser-ext/chrome/extension/annotations.js (hunk 15 above)
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: Ok

File: client/browser-ext/chrome/extension/inject.js (hunk 26 above)
Comment (attfarhan) [inline client/browser-ext/chrome/extension/inject.js]: Yes.  If we leave the branch, the srclib-data-version endpoint will not necessarily return a 404 since we do have a srclib-data-version for that branch.  If we sub it with the commit ID, it will return a 404 and we don't allow broken annotations to get injected.

File: client/browser-ext/chrome/extension/annotations.js (hunk 18 above)
Comment (attfarhan) [inline client/browser-ext/chrome/extension/annotations.js]: It actually does get reached when this fetch returns a 404.  Also, when I remove this, I get a bunch of unresolved promise errors.

File: client/browser-ext/chrome/extension/inject.js (hunk 27 above)
Comment (attfarhan) [inline client/browser-ext/chrome/extension/inject.js]: Yeah, this makes sense.  Made the switch, but to `if (repo !== this.props.repo) rev = defaultBranchCache[repo] || "master"`

File: client/browser-ext/chrome/extension/annotations.js [hunk 29]
```diff
@@ -140,20 +159,46 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+	let annRepo = [annURLsplit[1], annURLsplit[2], annURLsplit[3]]
+	let repo = annRepo.join("/");
+	if (fetchingDefaultBranchCache[repo]) {
+		return;
+	}
+	if (!defaultBranchCache[repo]) {
+		fetchingDefaultBranchCache[repo] = true;
+		fetch(`https://sourcegraph.com/.api/repos/${repo}`)
+			.then((response) => {
+				if (response.ok) {
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: you can condense the if/else block into the following:

```
defaultBranchCache[repo] = response.ok ? response.DefaultBranch : "master";
fetchingDefaultBranchCache[repo] = false;
```

^^ cleaner amirite? :)

File: client/browser-ext/chrome/extension/annotations.js [hunk 30]
```diff
@@ -140,20 +159,46 @@ function next(c, byteCount, annsByStartByte, annsByEndByte) {
...
+					defaultBranchCache[repo] = response.DefaultBranch;
+					fetchingDefaultBranchCache[repo] = false;
+				} else {
+					// Fall back onto master if we can't get the default branch
+					defaultBranchCache[repo] = "master";
+					fetchingDefaultBranchCache[repo] = false;
+				}
+			})
+			.catch((err) => console.log("Error getting default branch"))
```
Comment (rothfels) [inline client/browser-ext/chrome/extension/annotations.js]: This is still unreachable; you should remove it.

Comment (attfarhan) [inline client/browser-ext/chrome/extension/annotations.js]: @rothfels it does get reached when the fetch returns a 404. (screenshot: http://cl.ly/1d3N1y0W1G0R from here: https://github.com/alexo/wro4j/blob/1.7.x/wro4j-core/src/main/java/ro/isdc/wro/http/WroFilter.java) 

When I remove it, I get a bunch of unhandled promise rejections.

**Comment by rothfels** [comment]
@attfarhan this LGTM, can you have @uforic take a final QA pass before we merge this?

**Comment by uforic** [comment]
Went through test plan, tried out this morning in a variety of use cases, LGTM.
